pkg/kubelet/nodestatus/setters.go

```diff
@@ -73,16 +73,8 @@ func NodeAddress(nodeIP net.IP, // typically Kubelet.nodeIP
 			klog.V(2).Infof("Using node IP: %q", nodeIP.String())
 		}
 
-		if externalCloudProvider {
-			if nodeIP != nil {
-				if node.ObjectMeta.Annotations == nil {
-					node.ObjectMeta.Annotations = make(map[string]string)
-				}
-				node.ObjectMeta.Annotations[kubeletapis.AnnotationProvidedIPAddr] = nodeIP.String()
-			}
-			// We rely on the external cloud provider to supply the addresses.
-			return nil
-		}
+		// cloud provider is set
+		// try to discover it's node address from the cloud provider
 		if cloud != nil {
```

Comment:
is cloud `nil` when we have an external cloud provider?

Reply:
It comes from https://github.com/kubernetes/cloud-provider/blob/master/plugins.go#L109, so yes. Occasional think we would be better of with an explicit ExternalCloudProvider implementation but the current mechanisms (KCM, Kubelet, ...) have cloud set to nil when external.